addons/survey/static/src/js/survey_form.js

```diff
@@ -133,22 +145,134 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
             $("html, body").animate({ scrollTop: 0 }, "fast");
         }
         else if (result && result.fields && result.error === 'validation') {
-            var fieldKeys = _.keys(result.fields);
-            _.each(fieldKeys, function (key) {
-                self.$("#" + key + '>.o_survey_question_error').append($('<p>', {text: result.fields[key]})).removeClass('d-none');
-                if (fieldKeys[fieldKeys.length - 1] === key) {
-                    self._scrollToError(self.$('.o_survey_question_error:visible:first').closest('.js_question-wrapper'));
-                }
-            });
+            self.$('.o_survey_form_content').fadeIn(400);
+            self._showErrors(result.fields);
             return false;
         } else {
             var $target = self.$('.o_survey_error');
-            $target.removeClass('d-none');
+            $target.addClass("slide_in");
             self._scrollToError($target);
             return false;
         }
     },
 
+    // VALIDATION TOOLS
+    // -------------------------------------------------------------------------
+    /**
+    * Validation is done in frontend before submit to avoid latency from the server.
+    * If the validation is incorrect, the errors are displayed before submitting and
+    * fade in / out is avoided.
+    * @private
+    */
+    _validateForm: function ($form, formData) {
+        var self = this;
+        var errors = {};
+        var validationErrorMsg = _t("The answer you entered is not valid.");
+        var validationEmailMsg = _t("This answer must be an email address.");
+        var validationDateMsg = _t("This is not a date");
+        var constrErrorMsg = _t("This question requires an answer.");
+
+        this._resetErrors();
+
+        var data = {};
+        formData.forEach(function (value, key) {
+            data[key] = value;
+        });
+
+        $form.find('[data-question-type]').each(function () {
+            var $input = $(this);
+            var $questionWrapper = $input.closest(".js_question-wrapper");
+            var questionId = $questionWrapper[0].id;
+            var questionRequired = $questionWrapper.data('required');
+            switch ($input.data('question-type')) {
+                case 'char_box':
+                    if (questionRequired && !data[questionId]) {
```

Comment:
It's a little strange to test `!data[questionId]` where you could simply do `!$input.val()`.
I think it makes it easier to read as well.